### PR TITLE
add support for the kube-vip flavor

### PR DESCRIPTION
**Pull request by yastij**
Signed-off-by: Yassine TIJANI <ytijani@vmware.com>

**What this PR does / why we need it**: This PR adds support for a kube-vip flavor, two env variables are introduced:

`CONTROL_PLANE_ENDPOINT_IP`: the VIP we want to use
`VIP_NETWORK_INTERFACE`: the network interface we want to bind the VIP to (e.g. `eth0`)

**Which issue(s) this PR fixes**: Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
add support for a kube-vip flavor
```

### CODE REVIEW

**Comment by randomvariable** [comment]
PR needs some revisions to handle envsubst.
Closing and will resubmit.
/close

**Comment by yastij** [comment]
Rebased to fix CI

/assign @ncdc @detiber

File: packaging/flavorgen/flavors/generators.go [hunk 1]
```diff
@@ -396,6 +505,13 @@ func newHAProxyLoadBalancer() infrav1.HAProxyLoadBalancer {
 }
 
 func newKubeadmControlplane(replicas int, infraTemplate infrav1.VSphereMachineTemplate) controlplanev1.KubeadmControlPlane {
+	files := []bootstrapv1.File{
```
Comment (ncdc) [inline packaging/flavorgen/flavors/generators.go:508]: This should only be here if using kube-vip. It is currently also being generated for the HAProxy flavor.

**Comment by yastij** [comment]
@ncdc - pushed

File: Makefile [hunk 2]
```diff
@@ -116,7 +116,7 @@ e2e: e2e-image
 e2e: $(GINKGO) $(KUSTOMIZE) $(KIND) $(GOVC) ## Run e2e tests
 	$(MAKE) release-manifests
 	@mkdir -p $(E2E_TEMPLATE_DIR)
-	cp $(RELEASE_DIR)/cluster-template.yaml $(E2E_TEMPLATE_DIR)
+	cp $(RELEASE_DIR)/cluster-template-haproxy.yaml $(E2E_TEMPLATE_DIR)
```
Comment (ncdc) [inline Makefile:119]: When would you want to switch e2e to use kube-vip?

File: packaging/flavorgen/flavors/generators.go [hunk 3]
```diff
@@ -331,6 +341,105 @@ func defaultPreKubeadmCommands() []string {
...
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kube-vip",
+			Namespace: "kube-system",
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "kube-vip",
+					Image: "plndr/kube-vip:0.1.6",
```
Comment (ncdc) [inline packaging/flavorgen/flavors/generators.go:359]: Either here or a follow-up: let's allow the image to be customizable + defaulted if the user doesn't specify an alternative.

File: packaging/flavorgen/flavors/generators.go [hunk 4]
```diff
@@ -331,6 +341,105 @@ func defaultPreKubeadmCommands() []string {
...
+							Name:  "lb_backendport",
+							Value: "6443",
+						},
+						{
+							Name:  "vip_addpeerstolb",
+							Value: "true",
+						},
+						{
+							Name:  "lb_name",
+							Value: "kcpEdnpoint",
```
Comment (ncdc) [inline packaging/flavorgen/flavors/generators.go:401]:
```suggestion
							Value: "kcpEndpoint",
```

File: packaging/flavorgen/flavors/generators.go [hunk 5]
```diff
@@ -331,6 +341,105 @@ func defaultPreKubeadmCommands() []string {
...
+							Value: vipNetworkInterfaceVar,
+						},
+						{
+							Name:  "lb_type",
+							Value: "tcp",
+						},
+						{
+							Name:  "lb_type",
+							Value: "tcp",
+						},
```
Comment (ncdc) [inline packaging/flavorgen/flavors/generators.go:418]: Dupe of previous var

File: packaging/flavorgen/flavors/generators.go [hunk 6]
```diff
@@ -331,6 +341,105 @@ func defaultPreKubeadmCommands() []string {
...
+				},
+			},
+			HostNetwork: true,
+			Volumes: []v1.Volume{
+				{
+					Name: "kubeconfig",
+					VolumeSource: v1.VolumeSource{
+						HostPath: &v1.HostPathVolumeSource{
+							Path: "/etc/kubernetes/admin.conf",
+							Type: &hostPathType,
```
Comment (ncdc) [inline packaging/flavorgen/flavors/generators.go:429]: Let's use `pointer.StringPtr()`

File: Makefile (hunk 2 above)
Comment (yastij) [reply Makefile:119]: Ideally once DHCP is supported in kube-vip, or when doing some testing around VIPs from other segments in our infra.

File: packaging/flavorgen/flavors/generators.go (hunk 3 above)
Comment (yastij) [reply packaging/flavorgen/flavors/generators.go:359]: we don't enable custom images in our upstream templates (e.g. CCM and CSI), usually users have template overrides with their own images

File: packaging/flavorgen/flavors/generators.go (hunk 6 above)
Comment (yastij) [reply packaging/flavorgen/flavors/generators.go:429]: `pointer.StringPtr()` would return a `*string`, here we need a `*HostPathType, I'm not sure we can cast this ? did you have something specific in mind ?

Comment (ncdc) [reply packaging/flavorgen/flavors/generators.go:425]: Sorry, assumed it was a *string.

File: packaging/flavorgen/flavors/generators.go (hunk 3 above)
Comment (ncdc) [reply packaging/flavorgen/flavors/generators.go:359]: 👍

**Pull request merged**